README.md
# TradingReminder
Reminder for permitted to sell stocks, min value and amount


history.txt
            Action            |             Date             |           Quantity           |            Ticker            |            Value             
             buy              |          08/14/2017          |              1               |             GRPN             |            3.98
             buy              |          08/31/2017          |              19              |             DIS              |            101.7
             buy              |          09/05/2017          |              10              |             DIS              |            101.2
             buy              |          09/06/2017          |              4               |             DIS              |            101.0
             buy              |          09/06/2017          |              6               |             DIS              |            100.9
             buy              |          09/07/2017          |              10              |             DIS              |            100.81
             buy              |          09/07/2017          |              10              |             DIS              |            100.71
             buy              |          09/07/2017          |              10              |             DIS              |             98.2
             sell             |          09/08/2017          |              15              |             DIS              |            97.29
             buy              |          09/08/2017          |              10              |             DIS              |            96.75
             buy              |          09/12/2017          |              30              |             DIS              |             97.2
             sell             |          09/14/2017          |              35              |             DIS              |            98.66
             sell             |          09/14/2017          |              19              |             DIS              |            98.55
             buy              |          09/14/2017          |              15              |             DIS              |             98.3
             sell             |          09/14/2017          |              20              |             DIS              |            98.28
             buy              |          09/14/2017          |              25              |             DIS              |            98.02
             buy              |          09/14/2017          |              20              |             DIS              |            97.91
             buy              |          09/18/2017          |              15              |             XIV              |            93.18
             sell             |          09/18/2017          |              15              |             XIV              |            92.66
             buy              |          09/19/2017          |              60              |             ATVI             |            64.44
             buy              |          09/20/2017          |              25              |             AAPL             |            156.2
             buy              |          09/20/2017          |               8              |             AAPL             |            155.2
             buy              |          09/21/2017          |              10              |             LRCX             |            173.42
             buy              |          09/21/2017          |              15              |             LRCX             |            172.0
             buy              |          09/21/2017          |              20              |             AAPL             |            153.2
             buy              |          09/21/2017          |              14              |             NVDA             |            179.7
             buy              |          09/26/2017          |               6              |             NVDA             |            177.38
             buy              |          09/26/2017          |              15              |             NFLX             |            180.31
             buy              |          09/26/2017          |               6              |             BABA             |            170.34
             buy              |          09/26/2017          |               5              |             PYPL             |             63.5
             buy              |          09/28/2017          |              30              |             AMAT             |             49.35
             buy              |          10/02/2017          |              10              |             NFLX             |            178.0
             sell             |          10/02/2017          |              10              |             NFLX             |            178.0
             buy              |          10/04/2017          |              15              |             MSFT             |             74.0
             sell             |          10/04/2017          |              33              |             AAPL             |            153.50
             buy              |          10/04/2017          |             100              |             ROKU             |             22.00
             buy              |          10/04/2017          |              10              |             AAPL             |            152.80
             sell             |          10/05/2017          |              14              |             NVDA             |            180.16
             sell             |          10/05/2017          |              30              |             ROKU             |             22.03
             sell             |          10/05/2017          |              70              |             ROKU             |             22.02
             buy              |          10/09/2017          |               1              |             AMZN             |            996.95

main.py
#!/usr/bin/python3
import sys
import parser
import writehistory
import histreader
import policy
import color as Color
import util
from datetime import datetime
from collections import OrderedDict

columnWidth = 30
defaultHistoryFile = "history.txt"

def updateHistory(**kargs):
    print("updating...", kargs)

    if "fromCmd" in kargs:
        args = kargs["fromCmd"]
    else:
        args = kargs

    histFilename = defaultHistoryFile if not "historyFile" in args else args["historyFile"]
    outputStream = sys.stdout if not "outputStream" in args else args["outputStream"]
    printOptions = {
        "file": outputStream,
        "end": "<br>" if "outputStream" in args else "\n"
    }

    action, date, quantity, ticker, value = args["action"], args["date"], args["quantity"], args["ticker"], args["value"]
    writehistory.writeDataRow(histFilename, Action=action, Date=date, Quantity=quantity, Ticker=ticker, Value=value)

    print("Success!", printOptions)

def showReminder(**kargs):
    print("reminding...", kargs)
    print("\n\n")

    if "fromCmd" in kargs:
        args = kargs["fromCmd"]
        if "detail" in args and not args["detail"]:
            del args["detail"]
    else:
        args = kargs

    histFilename = defaultHistoryFile if not "historyFile" in args else args["historyFile"]
    outputStream = sys.stdout if not "outputStream" in args else args["outputStream"]
    printOptions = {
        "file": outputStream,
        "end": "<br>" if "outputStream" in args else "\n"
    }
    color = Color.disabled() if "outputStream" in args else Color
    trades = histreader.readHistoryFile(histFilename)
    result = policy.getSellLimit(trades, policy.processBloombergPolicy)

    numRecords = len(result["no sell limit"])

    result["no sell limit"] = OrderedDict(sorted(result["no sell limit"].items(), key=lambda i:i[0]))
    for idx, item in enumerate(result["no sell limit"].items()):
        if idx == 0:
            print("{0:=^50}".format(" No Sell Limit Holdings "), **printOptions)
        idxFormatStr = "{{0:>{}d}}".format(numRecords)
        print("    " , (idxFormatStr + ") Ticker: '" + color.R + "{1}" + color.W + "',  Quantity: " + color.Y + "{2}" + color.W).format(idx, item[0], item[1]), **printOptions)

        if idx == numRecords-1:
            print(**printOptions)

    result["limited"] = OrderedDict(sorted(result["limited"].items(), key=lambda i:i[0]))
    for idx, item in enumerate(result["limited"].items()):
        numRecords = len(str(len(result["limited"])))
        if idx == 0:
            print("{0:=^50}".format(" Limited Holdings "), **printOptions)
        idxFormatStr = "{{0:<{}}}".format(numRecords+1) # + ")"
        print("    ", (idxFormatStr + " Ticker: '" + color.R + "{1}" + color.W + "':").format(str(idx)+")", item[0]), **printOptions)
        numRecords = len(item[1])
        for subidx, limitedItem in enumerate(reversed(item[1])):
            idxFormatStr = "{{0:>{}}}".format(numRecords + 1) # + "."
            printStr = "        " + (idxFormatStr + "  Sell Limit at: " + color.G + "{1}" + color.W + ", Quantity: " + color.Y + "{2}" + color.W).format("i"*(subidx+1)+".", limitedItem["limit"], limitedItem["quantity"])
            if "detail" in args:
                daysToLift = (limitedItem["date"] - datetime.now().date()).days + 30
                spaceStr = "{{0:>{}d}}".format(80 - len(printStr))
                printStr += (", " + spaceStr + " days left").format(daysToLift)
            print(printStr, **printOptions)

    print(**printOptions)

    if "detail" in args:
        # show full information
        trades = histreader.readHistoryFile(histFilename)
        totalInvest = util.calculateTotalInvest(trades)
        print("Total investment amount: ${}".format(totalInvest), **printOptions)

def listRecords(**kargs):
    print("listing...", kargs)

    if "fromCmd" in kargs:
        args = kargs["fromCmd"]
    else:
        args = kargs

    histFilename = defaultHistoryFile if not "historyFile" in args else args["historyFile"]
    outputStream = sys.stdout if not "outputStream" in args else args["outputStream"]
    printOptions = {
        "file": outputStream,
        "end": "<br>" if "outputStream" in args else "\n"
    }

    # for now just show readin object
    hist = histreader.readHistoryFile(histFilename, plaintext=True)
    title = []
    if len(hist) > 0:
        title = sorted(hist[0].keys())

    title.append("Total")

    # Print title with reserved width for the line number
    lineNumberWidth = len(str(len(hist) + 1)) + 1
    strFormatList = ["{{{}:^{}}}".format(idx, columnWidth + lineNumberWidth) if idx == 0 else "{{{}:^{}}}".format(idx, columnWidth) for idx in range(len(title))]
    print("|".join(strFormatList).format(*tuple(title)), **printOptions)

    # Print records
    for idx, item in enumerate(hist):
        strFormatList = ["{{{}:^{}}}".format(idx, columnWidth) for idx in range(len(title))]
        strValueList = [round(int(item["Quantity"])*float(item["Value"]), 2) if key == "Total" else item[key] for key in title]
        lineNumberStr = "{})".format(idx+1)
        lineNumberFormat = "{{0:>{}}}".format(lineNumberWidth)
        print(lineNumberFormat.format(lineNumberStr) + "|".join(strFormatList).format(*strValueList), **printOptions)

if __name__ == "__main__":
    sys.argv.pop(0)

    startupParms = parser.parseCmdArgs(sys.argv)

    cmdHandlers = {
        "u"     :  updateHistory,
        "update":  updateHistory,
        "r"     :  showReminder,
        "remind":  showReminder,
        "l"     :  listRecords,
        "list"  :  listRecords
    }

    if startupParms == None:
        print("Failed to parser arguments. Abort")
        sys.exit(-1)

    if startupParms.topic in cmdHandlers:
        cmdHandlers[startupParms.topic](fromCmd=vars(startupParms))

    else:
        print("Invalid startupParms:", startupParms)
        sys.exit(-2)



parser.py
import sys
import argparse


def parseCmdArgs(args):
    if len(args) == 0:
        print("Invalid arguments:", args)
        return

    parser = argparse.ArgumentParser()

    subparsers = parser.add_subparsers(dest="topic", help="sub-command help")

    # create a parser for 'Update' command
    parser_remind = subparsers.add_parser("update", aliases=["u"], help="The update mode. Insert new record.")
    parser_remind.add_argument(
        "-n",
        "--ticker",
        help="Ticker name",
        required=True)

    parser_remind.add_argument(
        "-v",
        "--value",
        type=float,
        required=True,
        help="Trade value")

    parser_remind.add_argument(
        "-m",
        "--amount",
        type=int,
        required=True,
        help="Trade amount")

    parser_remind.add_argument(
        "-t",
        "--tradetype",
        choices=["sell", "buy"],
        required=True,
        help="Trade type")

    parser_remind.add_argument(
        "-d",
        "--tradedate",
        required=True,
        help="Trade date")

    # create a parser for 'Remind' command
    parser_remind = subparsers.add_parser("remind", aliases=["r"], help="Show what stocks/equities and the amount are permitted to be sold.")
    parser_remind.add_argument(
        "-d",
        "--detail",
        help="Detail mode",
        action="store_true")
    #group = parser_remind.add_mutually_exclusive_group(required=True)

    # create a parser for 'List' command
    parser_list = subparsers.add_parser("list", aliases=["l"], help="Display current holdings")
    parser_list.add_argument(
        "-hf",
        "--history",
        help="History filename")


    return parser.parse_args(args)


if __name__ == "__main__":
    sys.argv.pop(0)
    print(parseCmdArgs(sys.argv))



policy.py
from collections import deque
from datetime import datetime

def getSellLimit(trades, func, **kw):
    pureBoughtHistory = func(trades)

    # TODO: handle kw parameters

    result = {
        "no sell limit":{}, # {ticker: quantity(int)}
        "limited": {}       # {ticker: deque( { limit: int, quantity: int, date: date} ) }
    }

    asOfDay = datetime.now().date() # Currently only handle current date

    for item in pureBoughtHistory.items():
        ticker, trades = item[0], item[1]
        for trade in reversed(trades):
            date = trade["Date"]
            ticker = trade["Ticker"]
            quantity = trade["Quantity"]
            value = trade["Value"]

            if (asOfDay - date).days > 30:
                # no limit
                if not ticker in result["no sell limit"]:
                    result["no sell limit"][ticker] = 0
                result["no sell limit"][ticker] += int(quantity)

            else:
                if not ticker in result["limited"]:
                    result["limited"][ticker] = deque()
                stack = result["limited"][ticker]
                stack.appendleft({"limit": value, "quantity": quantity, "date": date})
                result["limited"][ticker] = stack

    return result


def processBloombergPolicy(trades):
    boughtStocksToTrades, soldStocksToTrades = {}, {}

    for trade in trades:
        ticker, action = trade["Ticker"], trade["Action"]

        if action == "sell":
            if ticker in soldStocksToTrades:
                stack = soldStocksToTrades[ticker]
                stack.appendleft(trade)
                soldStocksToTrades[ticker] = stack
            else:
                stack = deque()
                stack.appendleft(trade)
                soldStocksToTrades[ticker] = stack
        elif action == "buy":
            if ticker in boughtStocksToTrades:
                stack = boughtStocksToTrades[ticker]
                stack.appendleft(trade)
                boughtStocksToTrades[ticker] = stack
            else:
                stack = deque()
                stack.appendleft(trade)
                boughtStocksToTrades[ticker] = stack


    for item in soldStocksToTrades.items():
        ticker, soldTrades = item[0], item[1]
        boughtTrades = boughtStocksToTrades[ticker]

        fifo_single_ticker(soldTrades, boughtTrades)

        boughtStocksToTrades[ticker] = boughtTrades

    return boughtStocksToTrades


#
# param soldTrades: deque
# param boughtTrades: deque
#
def fifo_single_ticker(soldTrades, boughtTrades):
    if len(soldTrades) == 0:
        return;

    soldTrade, boughtTrade = soldTrades.pop(), boughtTrades.pop()
    q_s, q_b = int(soldTrade["Quantity"]), int(boughtTrade["Quantity"])

    if q_s > q_b:
        soldTrade["Quantity"] = str(q_s - q_b)
        soldTrades.append(soldTrade)
    elif q_s < q_b:
        boughtTrade["Quantity"] = str(q_b - q_s)
        boughtTrades.append(boughtTrade)

    return fifo_single_ticker(soldTrades, boughtTrades)

if __name__ == "__main__":
    histReader = __import__("histreader")
    trades = histReader.readHistoryFile("history.txt")
    print(getSellLimit(trades, processBloombergPolicy))


histreader.py
from datetime import datetime

def readHistoryFile(filename, **kw):
    # Read in history.txt and insert into a list
    lines = [line.rstrip('\n') for line in open(filename)]


    # Remove title line
    title = [line.strip() for line in lines.pop(0).split("|")]

    # Transform to tuples
    def transform(lineStr):
        d = {}
        l = [line.strip() for line in lineStr.split("|")]
        for idx, item in enumerate(l):
            colName = title[idx]
            if "plaintext" in kw:
                d[title[idx]] = item
            else:
                d[title[idx]] = datetime.strptime(item, "%m/%d/%Y").date() if colName == "Date" else item
        return d


    objs = list(map(transform, lines))

    return objs;


if __name__ == "__main__":
    print(readHistoryFile("history.txt"))


server/server.py
import sys
sys.path.insert(0, "../")

import main
import os
from io import StringIO
from flask import request

from flask import Flask, url_for, send_from_directory
app = Flask(__name__)

cmdHandlers = {
    "u"     :  main.updateHistory,
    "update":  main.updateHistory,
    "r"     :  main.showReminder,
    "remind":  main.showReminder,
    "l"     :  main.listRecords,
    "list"  :  main.listRecords
}

@app.route("/favicon.ico")
def favicon():
    return send_from_directory(os.path.join(app.root_path, '../image'), 'terminal.ico')

@app.route("/apple-touch-icon.png")
def appicon():
    return send_from_directory(os.path.join(app.root_path, '../image'), 'terminal-icon.png')

@app.route("/list")
def listRecordsRP():

    output = StringIO()
    startupParms = { "historyFile": "../history.txt", "outputStream": output }

    cmdHandlers["list"](**startupParms)
    return output.getvalue()


@app.route("/remind")
def remindSellLimitRP():

    output = StringIO()
    startupParms = { "historyFile": "../history.txt", "outputStream": output }

    # requests arguments
    if request.args.get("d") != None:
        startupParms["detail"] = ""

    cmdHandlers["remind"](**startupParms)
    return output.getvalue()


@app.route("/update")
def updateHistoryRP():

    output = StringIO()
    startupParms = { "historyFile": "../history.txt", "outputStream": output }

    # request arguments
    startupParms["action"]   = request.args.get("a")
    startupParms["date"]     = request.args.get("d")
    startupParms["quantity"] = request.args.get("q")
    startupParms["ticker"]   = request.args.get("t")
    startupParms["value"]    = request.args.get("v")


    cmdHandlers["update"](**startupParms)
    return output.getvalue()


writehistory.py
columnWidth = 30

columnNames = [ "Action", "Date", "Quantity", "Ticker", "Value" ]

def writeTitle(filename, **kargs):
    width = columnWidth
    if columnWidth in kargs:
        width = kargs["columnWidth"]

    rowFormatVec = ["{{{}:^{}}}".format(i, width) for i in range(5)]
    title = "|".join(rowFormatVec).format(*columnNames)
    f = open(filename, "w")
    f.write(title + "\n")


def writeDataRow(filename, **kargs):
    for check in columnNames:
       if not check in kargs:
            raise Exception({"reason": "No {}".format(check)})

    width = columnWidth
    if columnWidth in kargs:
        width = kargs["columnWidth"]

    rowFormatVec = ["{{{}:^{}}}".format(i, width) for i in range(5)]
    rowData = "|".join(rowFormatVec).format(
        kargs["Action"], kargs["Date"], kargs["Quantity"], kargs["Ticker"], kargs["Value"])

    f = open(filename, "a")
    f.write(rowData + "\n")

def copyFrom(fromFilename, toFilename, **kargs):
    width = columnWidth
    if columnWidth in kargs:
        width = columnWidth
    histreader = __import__("histreader")
    history = histreader.readHistoryFile(fromFilename, plaintext=True)
    writeTitle(toFilename)
    for record in history:
        action, date, ticker, value, quantity = record["Action"], record["Date"], record["Ticker"], record["Value"], record["Quantity"]
        writeDataRow(toFilename, Action=action, Date=date, Ticker=ticker, Value=value, Quantity=quantity, columnWidth=width)


if __name__ == "__main__":
    #writeTitle("test.txt", columnWidth=30)
    #writeDataRow("test.txt", Action="sell", Date="08/21/2019", Ticker="DIX", Value=30.4, Quantity=19, columnWidth=30)
    copyFrom("history.txt", "test.txt")
